In [5]:
! pip install pandas

In [6]:
! pip install plotly

# load data into dataframe

In [7]:
import pandas as pd

In [8]:
historical_df = pd.read_csv("historical_data.csv")

In [9]:
historical_df.head()

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Timestamp IST,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,02-12-2024 22:50,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,02-12-2024 22:50,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,02-12-2024 22:50,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,02-12-2024 22:50,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,02-12-2024 22:50,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12


In [10]:
fear_greed_df = pd.read_csv("fear_greed_index.csv")

fear_greed_df

,timestamp,value,classification,date
0,1517463000,30,Fear,2018-02-01
1,1517549400,15,Extreme Fear,2018-02-02
2,1517635800,40,Fear,2018-02-03
3,1517722200,24,Extreme Fear,2018-02-04
4,1517808600,11,Extreme Fear,2018-02-05
...,...,...,...,...
2639,1745818200,54,Neutral,2025-04-28
2640,1745904600,60,Greed,2025-04-29
2641,1745991000,56,Greed,2025-04-30
2642,1746077400,53,Neutral,2025-05-01


#  data  cleaning

## historical_df

In [11]:
historical_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13979 entries, 0 to 13978
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Account           13979 non-null  object 
 1   Coin              13979 non-null  object 
 2   Execution Price   13979 non-null  float64
 3   Size Tokens       13979 non-null  float64
 4   Size USD          13979 non-null  float64
 5   Side              13979 non-null  object 
 6   Timestamp IST     13979 non-null  object 
 7   Start Position    13979 non-null  float64
 8   Direction         13979 non-null  object 
 9   Closed PnL        13979 non-null  float64
 10  Transaction Hash  13979 non-null  object 
 11  Order ID          13978 non-null  float64
 12  Crossed           13978 non-null  object 
 13  Fee               13978 non-null  float64
 14  Trade ID          13978 non-null  float64
 15  Timestamp         13978 non-null  float64
dtypes: float64(9), object(7)
memory usage: 1

In [12]:
historical_df.describe()

,Execution Price,Size Tokens,Size USD,Start Position,Closed PnL,Order ID,Fee,Trade ID,Timestamp
count,13979.000000,13979.000000,1.397900e+04,13979.000000,13979.000000,1.397800e+04,13978.000000,1.397800e+04,1.397800e+04
mean,57599.338082,715.909666,2.996708e+04,7877.595828,89.472450,8.339395e+10,5.417314,5.658257e+14,1.743875e+12
std,41814.232877,10001.793818,1.234841e+05,59830.736129,786.395108,9.819651e+09,20.338225,3.242463e+14,5.969973e+09
min,0.007412,0.000010,1.000000e-01,-1000000.000000,-29370.119800,3.808060e+10,0.000000,0.000000e+00,1.730000e+12
25%,137.390000,0.016000,5.020400e+02,-1.584930,0.000000,8.561083e+10,0.005121,2.880000e+14,1.740000e+12
50%,84500.000000,0.292430,2.264680e+03,28.997900,0.000000,8.698356e+10,0.264203,5.650000e+14,1.740000e+12
75%,87800.000000,13.995000,1.499954e+04,103.868890,6.602400,8.845127e+10,2.499850,8.500000e+14,1.750000e+12
max,95420.000000,682429.000000,3.921431e+06,600000.000000,27223.741500,9.005577e+10,837.471593,1.130000e+15,1.750000e+12


In [13]:
historical_df.describe(include='object')

,Account,Coin,Side,Timestamp IST,Direction,Transaction Hash,Crossed
count,13979,13979,13979,13979,13979,13979,13978
unique,3,29,2,528,9,3850,2
top,0x513b8629fe877bb581bf244e326a047b249c4ff1,BTC,BUY,14-02-2025 00:31,Open Long,0x00000000000000000000000000000000000000000000...,True
freq,11986,9173,7325,441,5188,440,9798


In [14]:
historical_df.shape

(13979, 16)

In [15]:
historical_df.size

223664

In [16]:
historical_df.columns

Index(['Account', 'Coin', 'Execution Price', 'Size Tokens', 'Size USD', 'Side',
       'Timestamp IST', 'Start Position', 'Direction', 'Closed PnL',
       'Transaction Hash', 'Order ID', 'Crossed', 'Fee', 'Trade ID',
       'Timestamp'],
      dtype='object')

### check duplicates

In [17]:
historical_df.duplicated().sum()

np.int64(0)

In [18]:
historical_df.drop_duplicates()

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Timestamp IST,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,02-12-2024 22:50,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,02-12-2024 22:50,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,02-12-2024 22:50,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,02-12-2024 22:50,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,02-12-2024 22:50,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13974,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,957.10,139554.75,BUY,01-05-2025 00:33,39470.380000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,32.097592,1.100000e+15,1.750000e+12
13975,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,274.50,40024.85,BUY,01-05-2025 00:33,40427.480000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,9.205714,2.740000e+14,1.750000e+12
13976,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,91.66,13364.94,BUY,01-05-2025 00:33,40701.980000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,3.073937,6.800000e+14,1.750000e+12
13977,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8200,13.71,1999.19,BUY,01-05-2025 00:33,40793.640000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,0.459814,2.500000e+14,1.750000e+12


### data types

In [19]:
historical_df.dtypes

,0
Account,object
Coin,object
Execution Price,float64
Size Tokens,float64
Size USD,float64
Side,object
Timestamp IST,object
Start Position,float64
Direction,object
Closed PnL,float64


In [20]:
historical_df.head()

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Timestamp IST,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,02-12-2024 22:50,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,02-12-2024 22:50,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,02-12-2024 22:50,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,02-12-2024 22:50,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,02-12-2024 22:50,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12


In [21]:
historical_df['Timestamp IST'] = pd.to_datetime(historical_df['Timestamp IST'],format='%d-%m-%Y %H:%M')

In [22]:
historical_df['Date'] = historical_df['Timestamp IST'].dt.date

In [23]:
historical_df['Date']

,Date
0,2024-12-02
1,2024-12-02
2,2024-12-02
3,2024-12-02
4,2024-12-02
...,...
13974,2025-05-01
13975,2025-05-01
13976,2025-05-01
13977,2025-05-01


In [24]:
historical_df["Date"] = pd.to_datetime(historical_df["Date"])

In [25]:
historical_df.dtypes

,0
Account,object
Coin,object
Execution Price,float64
Size Tokens,float64
Size USD,float64
Side,object
Timestamp IST,datetime64[ns]
Start Position,float64
Direction,object
Closed PnL,float64


In [26]:
historical_df.drop(['Timestamp IST'],axis=1,inplace=True)

In [27]:
historical_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13979 entries, 0 to 13978
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Account           13979 non-null  object        
 1   Coin              13979 non-null  object        
 2   Execution Price   13979 non-null  float64       
 3   Size Tokens       13979 non-null  float64       
 4   Size USD          13979 non-null  float64       
 5   Side              13979 non-null  object        
 6   Start Position    13979 non-null  float64       
 7   Direction         13979 non-null  object        
 8   Closed PnL        13979 non-null  float64       
 9   Transaction Hash  13979 non-null  object        
 10  Order ID          13978 non-null  float64       
 11  Crossed           13978 non-null  object        
 12  Fee               13978 non-null  float64       
 13  Trade ID          13978 non-null  float64       
 14  Timestamp         1397

### check null values

In [28]:
historical_df.isnull().sum()

,0
Account,0
Coin,0
Execution Price,0
Size Tokens,0
Size USD,0
Side,0
Start Position,0
Direction,0
Closed PnL,0
Transaction Hash,0


### case sensitive

In [29]:
historical_df.head()

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp,Date
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12,2024-12-02
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12,2024-12-02
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12,2024-12-02
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12,2024-12-02
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12,2024-12-02


In [30]:
historical_df['Side'].unique()

array(['BUY', 'SELL'], dtype=object)

In [31]:
historical_df['Direction'].unique()

array(['Buy', 'Sell', 'Open Long', 'Close Long', 'Spot Dust Conversion',
       'Open Short', 'Close Short', 'Long > Short', 'Short > Long'],
      dtype=object)

In [32]:
historical_df['Crossed'].unique()

array([True, False, nan], dtype=object)

## fear_greed_df

In [33]:
fear_greed_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2644 entries, 0 to 2643
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   timestamp       2644 non-null   int64 
 1   value           2644 non-null   int64 
 2   classification  2644 non-null   object
 3   date            2644 non-null   object
dtypes: int64(2), object(2)
memory usage: 82.8+ KB


### check duplicates

In [34]:
fear_greed_df.duplicated().sum()

np.int64(0)

In [35]:
fear_greed_df.drop_duplicates()

,timestamp,value,classification,date
0,1517463000,30,Fear,2018-02-01
1,1517549400,15,Extreme Fear,2018-02-02
2,1517635800,40,Fear,2018-02-03
3,1517722200,24,Extreme Fear,2018-02-04
4,1517808600,11,Extreme Fear,2018-02-05
...,...,...,...,...
2639,1745818200,54,Neutral,2025-04-28
2640,1745904600,60,Greed,2025-04-29
2641,1745991000,56,Greed,2025-04-30
2642,1746077400,53,Neutral,2025-05-01


### dtypes

In [36]:
fear_greed_df.dtypes

,0
timestamp,int64
value,int64
classification,object
date,object


In [37]:
fear_greed_df['date'] = pd.to_datetime(fear_greed_df['date'])

In [38]:
fear_greed_df['Date'] = fear_greed_df['date']

In [39]:
fear_greed_df.head(2)

,timestamp,value,classification,date,Date
0,1517463000,30,Fear,2018-02-01,2018-02-01
1,1517549400,15,Extreme Fear,2018-02-02,2018-02-02


In [40]:
fear_greed_df.drop(['date'],axis=1,inplace=True)

In [41]:
fear_greed_df.head(2)

,timestamp,value,classification,Date
0,1517463000,30,Fear,2018-02-01
1,1517549400,15,Extreme Fear,2018-02-02


In [42]:
fear_greed_df.dtypes

,0
timestamp,int64
value,int64
classification,object
Date,datetime64[ns]


### check null values

In [43]:
fear_greed_df.isna().sum()

,0
timestamp,0
value,0
classification,0
Date,0


### case sensitive

In [44]:
fear_greed_df['classification'].unique()

array(['Fear', 'Extreme Fear', 'Neutral', 'Greed', 'Extreme Greed'],
      dtype=object)

# Merge two dataframe using date column

In [45]:
sentimental_analysis_df = pd.merge(
    historical_df,
    fear_greed_df,
    on="Date",
    how="left"
)

In [46]:
sentimental_analysis_df.head(3)

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp,Date,timestamp,value,classification
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed


In [47]:
sentimental_analysis_df.dtypes

,0
Account,object
Coin,object
Execution Price,float64
Size Tokens,float64
Size USD,float64
Side,object
Start Position,float64
Direction,object
Closed PnL,float64
Transaction Hash,object


In [48]:
sentimental_analysis_df.head()

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp,Date,timestamp,value,classification
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9769,986.87,7872.16,BUY,0.000000,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.345404,8.950000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed


## Handling null values after merge

In [49]:
sentimental_analysis_df.isna().sum()

,0
Account,0
Coin,0
Execution Price,0
Size Tokens,0
Size USD,0
Side,0
Start Position,0
Direction,0
Closed PnL,0
Transaction Hash,0


In [50]:
sentimental_analysis_df[sentimental_analysis_df.isnull().any(axis=1)]

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp,Date,timestamp,value,classification
727,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.8330,10000.00,28330.00,BUY,-32797.51,Close Short,11655.000000,0x4da9426d2bf6ec027d8404152f9bad02010700c745eb...,4.067141e+10,False,2.719680,8.350000e+13,1.730000e+12,2024-10-26,NaN,NaN,NaN
728,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.7110,10797.51,29272.05,BUY,-22797.51,Close Short,13901.794130,0x66588e180b942ff0e4d104152fb6f20000ac4ae48a39...,4.067144e+10,False,2.810115,3.230000e+14,1.730000e+12,2024-10-26,NaN,NaN,NaN
729,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.5889,6908.33,17884.98,BUY,-12000.00,Close Short,9737.981968,0x66588e180b942ff0e4d104152fb6f20000ac4ae48a39...,4.067147e+10,False,1.716957,1.070000e+15,1.730000e+12,2024-10-26,NaN,NaN,NaN
730,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.5889,529.39,1370.54,BUY,-5091.67,Close Short,746.228144,0x66588e180b942ff0e4d104152fb6f20000ac4ae48a39...,4.067147e+10,False,0.131570,8.770000e+14,1.730000e+12,2024-10-26,NaN,NaN,NaN
731,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.5889,38.65,100.06,BUY,-4562.28,Close Short,54.481040,0x66588e180b942ff0e4d104152fb6f20000ac4ae48a39...,4.067147e+10,False,0.009605,9.820000e+14,1.730000e+12,2024-10-26,NaN,NaN,NaN
732,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,EIGEN,2.5889,4523.63,11711.23,BUY,-4523.63,Close Short,6376.508848,0x39ce7efcc1773fd0eec704152fbcb4010100c8beadd4...,4.067147e+10,False,1.124277,9.360000e+14,1.730000e+12,2024-10-26,NaN,NaN,NaN
13978,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8200,6.92,1009.07,BUY,40807.35,Open Long,0.000000,0xcbb25369d6e8e8533f4604228e9bc001e7008cf,NaN,NaN,NaN,NaN,NaN,2025-05-01,1.746077e+09,53.0,Neutral


In [51]:
sentimental_analysis_df.shape

(13979, 19)

In [52]:
# check how much the data will be affected if the null values are dropped

6*100/211224

0.0028405862970117034

> Since only **0.0028 % ** of the dataset is affected, it is very lower, it doesn't affect the analysis, so these rows can be removed  

In [53]:
sentimental_analysis_df.dropna(subset=['timestamp','value','classification'],inplace=True)

In [54]:
sentimental_analysis_df.isna().sum()

,0
Account,0
Coin,0
Execution Price,0
Size Tokens,0
Size USD,0
Side,0
Start Position,0
Direction,0
Closed PnL,0
Transaction Hash,0


In [55]:
sentimental_analysis_df.to_csv("sentimental_analysis.csv",index=False)

# EDA & Data Visuvalization

In [56]:
import plotly.express as px

## 1. Sentiment Distribution

In [57]:
sentiment_count = (
    sentimental_analysis_df['classification']
    .value_counts()
    .reset_index()
)

sentiment_count

,classification,count
0,Fear,6535
1,Greed,3827
2,Neutral,2324
3,Extreme Fear,647
4,Extreme Greed,640


In [58]:
sentiment_count.columns = ['classification', 'count']

fig = px.bar(
    sentiment_count,
    x="classification",
    y="count",
    title="Market Sentiment Distribution",
    labels={
        "classification": "Sentiment",
        "count": "Number of Trades"
    },
    text="count",
    color="count",
    color_continuous_scale="Blues"
)

fig.update_traces(
    textposition='outside'
)

fig.show()

## Insights

### 1. Fear is the Most Common Market Sentiment
- Fear has the highest count (6,535).
- This indicates that the market was cautious for most of the observed period.

### 2. Greed is the Second Most Common Sentiment
- Greed appears 3,827 times in the dataset.
- Positive market sentiment was present but less frequent than Fear.

### 3. Extreme Market Conditions are Rare
- Extreme Fear (647) and Extreme Greed (640) have the lowest counts.
- Strong emotional market conditions occurred less frequently.

### 4. Negative Sentiment Dominates the Market
- Fear and Extreme Fear together account for 7,182 records.
- Greed and Extreme Greed together account for 4,467 records.
- This suggests that bearish sentiment was more common than bullish sentiment.

### 5. Most Market Conditions Fall Under Fear or Greed
- The majority of records belong to either Fear or Greed categories.
- These categories will be important when analyzing trader performance.

### 6. Extreme Fear and Extreme Greed May Indicate High Volatility
- Although less frequent, these periods may represent significant market movements.
- Further analysis can help determine how trader profits and losses behave during these conditions.

### Overall Insight
- The dataset shows that investors experienced Fear more often than Greed, indicating a generally cautious market environment during the analyzed period.

## 2. Average Profit by Sentiment

In [59]:
avg_pnl = (
    sentimental_analysis_df
    .groupby('classification')['Closed PnL']
    .mean()
    .reset_index()
)


In [60]:
avg_pnl

,classification,Closed PnL
0,Extreme Fear,198.456185
1,Extreme Greed,33.592341
2,Fear,79.542449
3,Greed,58.085638
4,Neutral,136.084118


In [61]:
fig = px.bar(
    avg_pnl,
    x="classification",
    y="Closed PnL",
    title="Average Profit by Market Sentiment",
    labels={
        "classification": "Sentiment",
        "Closed PnL": "Average Profit"
    },
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Greens"
)

fig.update_traces(
    texttemplate='%{text:.2f}',
    textposition='outside'
)

fig.show()

## Trader Performance by Market Sentiment

### 1. Highest Average Profit Occurred During Extreme Fear
- Traders achieved the highest average Closed PnL during Extreme Fear (198.46).
- This suggests that trading opportunities may be stronger when the market is highly pessimistic.

### 2. Neutral Market Conditions Also Show Strong Performance
- The average Closed PnL during Neutral sentiment is 136.08.
- Traders appear to perform well when the market is stable and less emotional.

### 3. Fear Outperforms Greed
- Average Closed PnL during Fear is 79.54, compared to 58.09 during Greed.
- Traders generated better profits in fearful market conditions than in optimistic ones.

### 4. Lowest Average Profit Occurred During Extreme Greed
- Extreme Greed has the lowest average Closed PnL at 33.59.
- This may indicate that traders face more challenges when the market becomes overly optimistic.

### 5. Contrarian Trading May Be Effective
- Profits are highest during Extreme Fear and lowest during Extreme Greed.
- This suggests that buying during market panic and avoiding excessive optimism could lead to better trading outcomes.

### Overall Insight
- Trader profitability appears to be stronger during fearful market conditions and weaker during highly greedy conditions. This indicates that market sentiment plays an important role in trading performance.

## 3. Total Profit by Sentiment

In [62]:
total_pnl = (
    sentimental_analysis_df
    .groupby('classification')['Closed PnL']
    .sum()
    .reset_index()
)


In [63]:
total_pnl

,classification,Closed PnL
0,Extreme Fear,128401.151696
1,Extreme Greed,21499.098510
2,Fear,519809.902183
3,Greed,222293.736806
4,Neutral,316259.490707


In [64]:
fig = px.bar(
    total_pnl,
    x="classification",
    y="Closed PnL",
    title="Total Profit by Market Sentiment",
    labels={
        "classification": "Sentiment",
        "Closed PnL": "Total Profit"
    },
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Viridis"
)

fig.update_traces(
    texttemplate='%{text:.2s}',
    textposition='outside'
)

fig.show()

## Total Closed PnL by Market Sentiment

### 1. Fear Generated the Highest Total Profit
- Fear recorded the highest total Closed PnL of 519,809.90.
- This indicates that traders earned the most profits during fearful market conditions.

### 2. Neutral Sentiment Shows Strong Profitability
- Neutral sentiment generated a total Closed PnL of 316,259.49.
- Traders were able to achieve consistent profits even when the market was neither fearful nor greedy.

### 3. Greed Produced Moderate Profits
- Greed resulted in a total Closed PnL of 222,293.74.
- While profitable, it performed lower than both Fear and Neutral market conditions.

### 4. Extreme Fear Outperformed Extreme Greed
- Extreme Fear generated 128,401.15 in total Closed PnL.
- Extreme Greed generated only 21,499.10 in total Closed PnL.
- Traders earned significantly higher profits during market panic than during excessive optimism.

### 5. Market Fear Appears More Favorable for Traders
- Both average and total profitability are stronger during Fear-related sentiments.
- This suggests that traders may find better opportunities when the market is pessimistic.

### Overall Insight
- The highest trading profits were achieved during Fear and Neutral market conditions, while Extreme Greed delivered the lowest profitability. This indicates that trader performance tends to improve during cautious or uncertain market environments rather than highly optimistic market phases.

## 4. Win Rate by Sentiment

In [65]:
sentimental_analysis_df['Win'] = (
    sentimental_analysis_df['Closed PnL'] > 0
).astype(int)

win_rate = (
    sentimental_analysis_df
    .groupby('classification')['Win']
    .mean()
    .reset_index()
)


In [66]:
win_rate

,classification,Win
0,Extreme Fear,0.315301
1,Extreme Greed,0.165625
2,Fear,0.388217
3,Greed,0.347269
4,Neutral,0.498709


In [67]:
win_rate['Win'] = win_rate['Win'] * 100

In [68]:
win_rate

,classification,Win
0,Extreme Fear,31.530139
1,Extreme Greed,16.562500
2,Fear,38.821729
3,Greed,34.726940
4,Neutral,49.870912


In [69]:
win_rate['Win'] = win_rate['Win'] * 100

fig = px.bar(
    win_rate,
    x="classification",
    y="Win",
    title="Win Rate by Market Sentiment",
    labels={
        "classification": "Sentiment",
        "Win": "Win Rate (%)"
    },
    text="Win",
    color="Win",
    color_continuous_scale="Teal"
)

fig.update_traces(
    texttemplate='%{text:.2f}%',
    textposition='outside'
)

fig.show()

## Winning Trades by Market Sentiment

### 1. Neutral Sentiment Has the Highest Number of Winning Trades
- Neutral sentiment recorded the highest winning trades value (4,987.09).
- Traders appear to perform best when the market is stable and less driven by emotions.

### 2. Fear Shows Strong Trading Success
- Fear recorded 3,882.17 winning trades.
- This suggests that traders can still achieve strong results during cautious market conditions.

### 3. Greed Also Generates Many Winning Trades
- Greed recorded 3,472.69 winning trades.
- Positive market sentiment provides profitable opportunities, but not as many as Neutral or Fear conditions.

### 4. Extreme Fear Outperforms Extreme Greed
- Extreme Fear recorded 3,153.01 winning trades.
- Extreme Greed recorded only 1,656.25 winning trades.
- Traders seem to achieve better results during market panic than during excessive optimism.

### 5. Extreme Greed Has the Lowest Trading Success
- Extreme Greed has the lowest winning trades value among all sentiment categories.
- This may indicate that overconfidence in the market can reduce trading performance.

### Overall Insight
- Winning trades are highest during Neutral and Fear market conditions, while Extreme Greed shows the weakest performance. This suggests that traders tend to achieve better results in balanced or cautious markets than in highly optimistic market environments.

5. Average Trade Size

In [70]:
trade_size = (
    sentimental_analysis_df
    .groupby('classification')['Size USD']
    .mean()
    .reset_index()
)

In [71]:
trade_size

,classification,Size USD
0,Extreme Fear,6539.688454
1,Extreme Greed,3061.835375
2,Fear,34584.931080
3,Greed,32256.562869
4,Neutral,27182.414781


In [72]:
fig = px.bar(
    trade_size,
    x="classification",
    y="Size USD",
    title="Average Trade Size by Sentiment",
    labels={
        "classification": "Sentiment",
        "Size USD": "Average Trade Size (USD)"
    },
    text="Size USD",
    color="Size USD",
    color_continuous_scale="Purples"
)

fig.update_traces(
    texttemplate='%{text:.2s}',
    textposition='outside'
)

fig.show()

## Trading Volume (Size USD) by Market Sentiment

### 1. Fear Has the Highest Trading Volume
- Fear recorded the highest total trading volume of 34,584.93 USD.
- Traders were most active during fearful market conditions.

### 2. Greed Also Shows High Trading Activity
- Greed generated a trading volume of 32,256.56 USD.
- This indicates strong market participation during optimistic periods.

### 3. Neutral Sentiment Maintains Significant Activity
- Neutral sentiment recorded 27,182.41 USD in trading volume.
- Traders remained actively involved even when the market was stable.

### 4. Extreme Fear Generates More Volume Than Extreme Greed
- Extreme Fear recorded 6,539.69 USD in trading volume.
- Extreme Greed recorded only 3,061.84 USD.
- Traders appear to participate more during market panic than during excessive optimism.

### 5. Trading Activity Increases During Emotional Market Conditions
- Fear and Greed account for the majority of trading volume.
- This suggests that market emotions play an important role in driving trading activity.

### Overall Insight
- The highest trading volume was observed during Fear and Greed periods, with Fear leading all sentiment categories. This indicates that traders are more active when market emotions are strong, particularly during fearful conditions.

## 6. Box Plot (Most Important)

In [73]:
fig = px.box(
    sentimental_analysis_df,
    x="classification",
    y="Closed PnL",
    color="classification",
    title="PnL Distribution by Market Sentiment"
)

fig.show()

- The box plot shows that trader profitability varies across sentiment, with higher spread and outliers during emotional market conditions like Fear and Greed, indicating increased volatility and trading opportunities

## 7. Correlation Heatmap

In [74]:
corr_cols = [
    'Execution Price',
    'Size Tokens',
    'Size USD',
    'Closed PnL',
    'Fee',
    'value'
]

corr_matrix = sentimental_analysis_df[corr_cols].corr()



In [75]:
corr_matrix

,Execution Price,Size Tokens,Size USD,Closed PnL,Fee,value
Execution Price,1.000000,-0.098302,0.136725,-0.059749,0.143460,-0.126533
Size Tokens,-0.098302,1.000000,-0.011289,0.045085,-0.009144,-0.000096
Size USD,0.136725,-0.011289,1.000000,0.323486,0.707993,-0.012767
Closed PnL,-0.059749,0.045085,0.323486,1.000000,0.198181,-0.038890
Fee,0.143460,-0.009144,0.707993,0.198181,1.000000,0.008034
value,-0.126533,-0.000096,-0.012767,-0.038890,0.008034,1.000000


In [76]:
fig = px.imshow(
    corr_matrix,
    text_auto=True,
    title="Correlation Heatmap",
    color_continuous_scale="RdBu"
)

fig.show()

## Correlation Analysis Insights

### 1. Trade Size (USD) Has the Strongest Relationship with Fees
- Size USD and Fee have a strong positive correlation (0.71).
- Larger trades generally result in higher trading fees.

### 2. Trade Size (USD) Positively Impacts Profitability
- Size USD and Closed PnL show a moderate positive correlation (0.32).
- Traders placing larger trades tend to generate higher profits.

### 3. Fees Increase with Trade Profitability
- Fee and Closed PnL have a weak positive correlation (0.20).
- Profitable trades are often associated with higher fees because they usually involve larger trade sizes.

### 4. Execution Price Has Little Impact on Profitability
- Execution Price and Closed PnL have a very weak negative correlation (-0.06).
- Entry price alone does not significantly influence trading profits.

### 5. Size Tokens Has Almost No Relationship with Other Variables
- Size Tokens shows correlations close to zero with most features.
- The number of tokens traded does not appear to strongly affect profit, fees, or trade value.

### 6. No Strong Negative Correlations Found
- Most correlations are weak or positive.
- There is no evidence of a strong inverse relationship between the analyzed trading metrics.

### Key Business Insight
- Trade size is the most important factor influencing trader performance. Larger trades are associated with higher profits, but they also lead to higher trading fees. Traders should focus on position sizing strategies to maximize profitability while controlling transaction costs.

# Do traders earn profit during Fear or Greed?

In [77]:
profit_by_sentiment = (
    sentimental_analysis_df
    .groupby('classification')['Closed PnL']
    .mean()
    .reset_index()
)

In [79]:
profit_by_sentiment

,classification,Closed PnL
0,Extreme Fear,198.456185
1,Extreme Greed,33.592341
2,Fear,79.542449
3,Greed,58.085638
4,Neutral,136.084118


In [78]:
fig = px.bar(
    profit_by_sentiment,
    x="classification",
    y="Closed PnL",
    title="Average Profit by Market Sentiment",
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Blues"
)

fig.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig.show()

## PnL Distribution by Market Sentiment (Box Plot Insight)

### 1. Shows Profit/Loss Spread for Each Sentiment
- This plot compares how trader profits (Closed PnL) are distributed across different market sentiments.

### 2. Helps Identify Best and Worst Market Conditions
- You can clearly see which sentiment gives higher profits and which gives losses.

### 3. Median Profit Comparison
- The middle line in each box shows the median profit.
- This helps understand the “typical” trader performance in each sentiment.

### 4. Extreme Values (Outliers)
- Points outside the box show very high profits or losses.
- These represent rare but important trading outcomes.

### 5. Risk vs Reward Insight
- Wide box = high volatility (riskier trading conditions)
- Narrow box = stable performance

### 6. Business Insight
- If Fear or Neutral shows higher median PnL → traders perform better in cautious/normal markets.
- If Greed or Extreme Greed shows lower median → overconfidence may reduce profits.

# Do traders use higher leverage in Greed market?

In [83]:
leverage_by_sentiment = (
    sentimental_analysis_df
    .groupby('classification')['Start Position']
    .mean()
    .reset_index()
)

In [85]:
leverage_by_sentiment

,classification,Start Position
0,Extreme Fear,16990.582744
1,Extreme Greed,2398.942883
2,Fear,5416.268693
3,Greed,11303.630871
4,Neutral,8184.226737


In [81]:
fig = px.bar(
    leverage_by_sentiment,
    x="classification",
    y="Start Position",
    title="Average Leverage by Market Sentiment",
    text="Start Position",
    color="Start Position",
    color_continuous_scale="Reds"
)

fig.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig.show()

## Average Start Position by Market Sentiment

### 1. Extreme Fear Has the Highest Average Start Position
- Extreme Fear recorded the highest average start position (16,990.58).
- Traders tend to open larger positions when the market is experiencing extreme fear.

### 2. Greed Shows the Second Highest Position Size
- Greed has an average start position of 11,303.63.
- Traders appear more confident and take larger positions during optimistic market conditions.

### 3. Neutral Market Conditions Show Moderate Position Sizes
- Neutral sentiment recorded an average start position of 8,184.23.
- Trading activity remains balanced when market sentiment is stable.

### 4. Fear Has Smaller Average Positions Than Expected
- Fear recorded an average start position of 5,416.27.
- While traders are active during Fear periods, they tend to take relatively smaller positions compared to Extreme Fear and Greed.

### 5. Extreme Greed Has the Lowest Average Start Position
- Extreme Greed recorded the lowest average start position (2,398.94).
- Traders may become more cautious or reduce exposure during excessively optimistic market conditions.

### Overall Insight
- Traders take the largest positions during Extreme Fear, suggesting that some traders view panic-driven markets as opportunities. In contrast, position sizes are smallest during Extreme Greed, indicating reduced risk-taking despite strong market optimism.

# Buy vs Sell performance in Fear & Greed

In [86]:
buy_sell_pnl = (
    sentimental_analysis_df
    .groupby(['classification', 'Side'])['Closed PnL']
    .mean()
    .reset_index()
)

In [87]:
buy_sell_pnl

,classification,Side,Closed PnL
0,Extreme Fear,BUY,385.713801
1,Extreme Fear,SELL,-112.869234
2,Extreme Greed,BUY,0.000000
3,Extreme Greed,SELL,67.820500
4,Fear,BUY,42.572095
5,Fear,SELL,120.008397
6,Greed,BUY,14.170337
7,Greed,SELL,110.426841
8,Neutral,BUY,0.000000
9,Neutral,SELL,257.540302


In [88]:
fig = px.bar(
    buy_sell_pnl,
    x="classification",
    y="Closed PnL",
    color="Side",
    barmode="group",
    title="Buy vs Sell Profit by Sentiment",
    text="Closed PnL"
)

fig.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig.show()

# What is Win Rate in Fear market?

In [89]:
sentimental_analysis_df['Win'] = (
    sentimental_analysis_df['Closed PnL'] > 0
).astype(int)

win_rate = (
    sentimental_analysis_df
    .groupby('classification')['Win']
    .mean()
    .reset_index()
)

win_rate['Win'] = win_rate['Win'] * 100


In [91]:
win_rate

,classification,Win
0,Extreme Fear,31.530139
1,Extreme Greed,16.562500
2,Fear,38.821729
3,Greed,34.726940
4,Neutral,49.870912


In [90]:
fig = px.bar(
    win_rate,
    x="classification",
    y="Win",
    title="Win Rate by Market Sentiment",
    text="Win",
    color="Win",
    color_continuous_scale="Greens"
)

fig.update_traces(texttemplate='%{text:.2f}%', textposition='outside')
fig.show()

## Win Rate Analysis by Market Sentiment

### 1. Neutral Sentiment Has the Highest Win Rate
- Neutral sentiment recorded the highest win rate (49.87%).
- Traders appear to perform best when the market is stable and less influenced by strong emotions.

### 2. Fear Shows Strong Trading Performance
- Fear achieved a win rate of 38.82%.
- Despite negative market sentiment, traders were able to maintain a relatively high success rate.

### 3. Greed Performs Slightly Lower Than Fear
- Greed recorded a win rate of 34.73%.
- Trader success was slightly lower during optimistic market conditions compared to fearful conditions.

### 4. Extreme Fear Outperforms Extreme Greed
- Extreme Fear achieved a win rate of 31.53%, while Extreme Greed achieved only 16.56%.
- Traders were nearly twice as successful during Extreme Fear compared to Extreme Greed.

### 5. Extreme Greed Has the Lowest Win Rate
- Extreme Greed recorded the lowest win rate among all sentiment categories.
- Excessive market optimism may lead to poorer trading decisions and lower success rates.

### Overall Insight
- Trader win rates are highest during Neutral and Fear market conditions, while Extreme Greed shows the weakest performance. This suggests that traders tend to make better decisions in stable or cautious markets than in highly optimistic market environments.

# High leverage + Greed = Loss risk

In [92]:
high_leverage = sentimental_analysis_df[
    sentimental_analysis_df['Start Position'] > sentimental_analysis_df['Start Position'].quantile(0.75)
]



In [94]:
high_leverage

,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp,Date,timestamp,value,classification,Win
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9800,16.00,127.68,BUY,986.524596,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.005600,4.430000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed,0
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9855,144.09,1150.63,BUY,1002.518996,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050431,6.600000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed,0
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9874,142.98,1142.04,BUY,1146.558564,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.050043,1.080000e+15,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed,0
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9894,8.73,69.75,BUY,1289.488521,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.003055,1.050000e+15,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed,0
5,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.9900,1.41,11.27,BUY,1298.215466,Buy,0.0,0xec09451986a1874e3a980418412fcd0201f500c95bac...,5.201771e+10,True,0.000493,7.330000e+14,1.730000e+12,2024-12-02,1.733117e+09,80.0,Extreme Greed,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13974,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,957.10,139554.75,BUY,39470.380000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,32.097592,1.100000e+15,1.750000e+12,2025-05-01,1.746077e+09,53.0,Neutral,0
13975,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,274.50,40024.85,BUY,40427.480000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,9.205714,2.740000e+14,1.750000e+12,2025-05-01,1.746077e+09,53.0,Neutral,0
13976,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8100,91.66,13364.94,BUY,40701.980000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,3.073937,6.800000e+14,1.750000e+12,2025-05-01,1.746077e+09,53.0,Neutral,0
13977,0x513b8629fe877bb581bf244e326a047b249c4ff1,SOL,145.8200,13.71,1999.19,BUY,40793.640000,Open Long,0.0,0xcbb25369d6e8e8533f4604228e9bc001e7008cf50e80...,9.005577e+10,True,0.459814,2.500000e+14,1.750000e+12,2025-05-01,1.746077e+09,53.0,Neutral,0


In [93]:
fig = px.box(
    high_leverage,
    x="classification",
    y="Closed PnL",
    title="High Leverage Trades PnL by Sentiment"
)

fig.show()

# Buy vs Sell strength in each sentiment

In [95]:
direction_profit = (
    sentimental_analysis_df
    .groupby(['classification', 'Side'])['Closed PnL']
    .sum()
    .reset_index()
)



In [97]:
direction_profit

,classification,Side,Closed PnL
0,Extreme Fear,BUY,155828.375546
1,Extreme Fear,SELL,-27427.223850
2,Extreme Greed,BUY,0.000000
3,Extreme Greed,SELL,21499.098510
4,Fear,BUY,145383.704372
5,Fear,SELL,374426.197811
6,Greed,BUY,29488.471790
7,Greed,SELL,192805.265016
8,Neutral,BUY,0.000000
9,Neutral,SELL,316259.490707


In [96]:
fig = px.bar(
    direction_profit,
    x="classification",
    y="Closed PnL",
    color="Side",
    barmode="group",
    title="Total Buy vs Sell Profit by Sentiment",
    text="Closed PnL"
)

fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.show()

## Total Closed PnL by Side and Market Sentiment

### 1. SELL Trades Generated Most of the Total Profit
- SELL trades produced higher total profits than BUY trades in Fear, Greed, and Neutral market conditions.
- This indicates that SELL strategies were generally more profitable during the analyzed period.

### 2. Fear + SELL Combination Delivered the Highest Profit
- SELL trades during Fear generated the highest total Closed PnL (374,426.20).
- Traders achieved their best overall performance in fearful market conditions using SELL positions.

### 3. Neutral Market Also Showed Strong SELL Performance
- SELL trades in Neutral sentiment generated 316,259.49 in total profit.
- Stable market conditions provided significant profitable trading opportunities.

### 4. BUY Trades Performed Exceptionally Well During Extreme Fear
- BUY trades generated 155,828.38 in total profit during Extreme Fear.
- This suggests that buying during market panic can be highly rewarding.

### 5. SELL Trades Lost Money During Extreme Fear
- SELL trades recorded a loss of -27,427.22 during Extreme Fear.
- Panic-driven market reversals may have negatively impacted short positions.

### 6. Extreme Greed and Neutral Profits Came Entirely from SELL Trades
- BUY trades generated no profit in Extreme Greed and Neutral sentiments.
- All recorded profits during these periods came from SELL trades.

### 7. Greed Markets Favored SELL Strategies
- During Greed, SELL trades generated 192,805.27 compared to only 29,488.47 from BUY trades.
- Taking profits or short-selling was considerably more effective than buying.

### Overall Insight
- SELL strategies contributed the majority of total trading profits across most market conditions. However, during Extreme Fear, BUY trades significantly outperformed SELL trades, suggesting that market panic may create strong buying opportunities while reducing the effectiveness of short positions.

# Which coin gives highest profit in Fear vs Greed?

In [98]:
coin_profit = (
    sentimental_analysis_df
    .groupby(['Coin', 'classification'])['Closed PnL']
    .sum()
    .reset_index()
)

top_coins = coin_profit.sort_values('Closed PnL', ascending=False).head(15)



In [99]:
top_coins

,Coin,classification,Closed PnL
16,BTC,Fear,322749.441392
1,@107,Neutral,194499.072534
32,HYPE,Extreme Fear,160623.306862
34,HYPE,Fear,118971.158972
18,BTC,Neutral,103285.886682
17,BTC,Greed,101835.738550
55,SOL,Greed,62207.716396
9,AIXBT,Greed,38168.354860
54,SOL,Fear,35384.971160
56,SOL,Neutral,19136.507338


In [100]:
fig = px.bar(
    top_coins,
    x="Coin",
    y="Closed PnL",
    color="classification",
    title="Top Coins Profit by Sentiment",
    text="Closed PnL",
    barmode="group"
)

fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.show()

## Top Coin Performance by Market Sentiment

### 1. BTC Generated the Highest Profit During Fear
- BTC recorded the highest Closed PnL of 322,749.44 during Fear sentiment.
- This indicates that Bitcoin trading was highly profitable during cautious market conditions.

### 2. The Highest Profit Outside BTC Came from @107 During Neutral Markets
- @107 generated 194,499.07 in Closed PnL during Neutral sentiment.
- Traders achieved strong results when market emotions were balanced.

### 3. HYPE Performed Exceptionally Well During Extreme Fear
- HYPE generated 160,623.31 in Closed PnL during Extreme Fear.
- This suggests that HYPE benefited significantly from panic-driven market conditions.

### 4. BTC Remained Profitable Across Multiple Sentiments
- BTC generated strong profits in Fear (322,749.44), Neutral (103,285.89), and Greed (101,835.74).
- This shows consistent trader profitability across different market environments.

### 5. Fear Sentiment Dominates Top Profit Opportunities
- Several top-performing combinations occurred during Fear sentiment, including BTC, HYPE, SOL, LAYER, and TRUMP.
- Fearful markets appear to provide more profitable trading opportunities.

### 6. SOL Performed Well Across Different Sentiments
- SOL generated profits during Fear, Greed, Neutral, and Extreme Fear.
- This indicates stable performance regardless of market sentiment.

### 7. Extreme Greed Appears Less Profitable
- Only one entry from Extreme Greed appears among the top performers (@107 with 12,793.91).
- This supports earlier findings that trader performance tends to weaken during highly optimistic market conditions.

### Overall Insight
- BTC was the most profitable asset, especially during Fear sentiment. Most top-performing coin-sentiment combinations occurred during Fear and Neutral market conditions, suggesting that traders achieved better results in cautious or balanced markets than in highly optimistic ones.

# Does high trade size increase profit?

In [102]:
size_profit = sentimental_analysis_df.copy()

size_profit['Size Bucket'] = pd.qcut(
    size_profit['Size USD'],
    5,
    labels=["Very Low", "Low", "Medium", "High", "Very High"]
)

size_analysis = (
    size_profit
    .groupby('Size Bucket')['Closed PnL']
    .mean()
    .reset_index()
)



/tmp/ipykernel_4716/1599652792.py:11: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [103]:
size_analysis

,Size Bucket,Closed PnL
0,Very Low,1.676501
1,Low,12.191785
2,Medium,73.748213
3,High,91.975172
4,Very High,252.863823


In [104]:
fig = px.bar(
    size_analysis,
    x="Size Bucket",
    y="Closed PnL",
    title="Trade Size vs Profit",
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Viridis"
)

fig.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig.show()

## Profitability by Trade Size

### 1. Profit Increases with Trade Size
- Average Closed PnL increases as trade size grows.
- Larger trades tend to generate higher profits than smaller trades.

### 2. Very Low Trade Sizes Generate Minimal Profit
- Very Low trade sizes recorded an average Closed PnL of only 1.68.
- Small positions contributed the least to overall profitability.

### 3. Medium and High Trade Sizes Show Strong Performance
- Medium-sized trades generated 73.75 in average profit.
- High-sized trades generated 91.98 in average profit.
- Increasing position size appears to improve profit potential.

### 4. Very High Trade Sizes Deliver the Highest Profit
- Very High trade sizes recorded the highest average Closed PnL of 252.86.
- These trades generated nearly three times the profit of High-sized trades.

### 5. Clear Positive Relationship Between Trade Size and Profit
- Profitability consistently increases from Very Low to Very High trade sizes.
- This suggests that position size is a key factor influencing trader returns.

### Overall Insight
- Traders with larger position sizes achieved significantly higher profits than those with smaller positions. The results indicate a strong positive relationship between trade size and profitability, highlighting the importance of effective position sizing in trading performance.

# Does fee reduce profit significantly?

In [105]:
fee_analysis = (
    sentimental_analysis_df
    .groupby('classification')[['Fee', 'Closed PnL']]
    .mean()
    .reset_index()
)



In [107]:
fee_analysis


,classification,Fee,Closed PnL
0,Extreme Fear,1.082042,198.456185
1,Extreme Greed,0.413626,33.592341
2,Fear,5.777581,79.542449
3,Greed,6.561108,58.085638
4,Neutral,5.115820,136.084118


In [106]:
fig = px.scatter(
    fee_analysis,
    x="Fee",
    y="Closed PnL",
    color="classification",
    size="Fee",
    title="Fee vs Profit by Sentiment"
)

fig.show()

## Fee vs Profit Analysis by Market Sentiment

### 1. Higher Fees Do Not Always Lead to Higher Profits
- Greed recorded the highest average fee (6.56), but the average profit was only 58.09.
- This suggests that paying higher trading fees does not necessarily result in better returns.

### 2. Extreme Fear Delivered the Highest Profit with Low Fees
- Extreme Fear generated the highest average Closed PnL (198.46) with a relatively low average fee (1.08).
- Traders achieved strong profitability while incurring lower transaction costs.

### 3. Neutral Markets Show Strong Profit Efficiency
- Neutral sentiment recorded an average profit of 136.08 with a fee of 5.12.
- Traders were able to maintain high profitability without the highest fee levels.

### 4. Extreme Greed Had Both Low Fees and Low Profits
- Extreme Greed recorded the lowest average fee (0.41) and the lowest average profit (33.59).
- Lower trading costs alone were not enough to improve performance.

### 5. Fear Outperformed Greed Despite Lower Fees
- Fear generated higher profits (79.54) than Greed (58.09) while having lower average fees.
- This indicates that market conditions may have a stronger impact on profitability than trading costs.

### Overall Insight
- Market sentiment appears to influence profitability more than trading fees. The highest profits were observed during Extreme Fear and Neutral conditions, while Greed and Extreme Greed produced lower returns despite comparable or lower transaction costs.

# Do experienced traders perform better?

In [108]:
account_profit = (
    sentimental_analysis_df
    .groupby('Account')['Closed PnL']
    .sum()
    .reset_index()
)

top_accounts = account_profit.sort_values('Closed PnL', ascending=False).head(10)


In [111]:
top_accounts

,Account,Closed PnL
0,0x513b8629fe877bb581bf244e326a047b249c4ff1,779878.250342
1,0x72c6a4624e1dffa724e6d00d64ceae698af892a0,360539.510029
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,67845.619531


In [110]:
fig = px.bar(
    top_accounts,
    x="Account",
    y="Closed PnL",
    title="Top Performing Traders",
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Blues"
)

fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.show()

## Top Traders by Total Profit

### 1. One Trader Significantly Outperformed Others
- Account `0x513b8629fe877bb581bf244e326a047b249c4ff1` generated the highest total Closed PnL of 779,878.25.
- This trader earned more than twice the profit of the second-ranked trader.

### 2. Strong Gap Between First and Second Place
- The top trader earned 779,878.25, while the second trader earned 360,539.51.
- This large difference suggests exceptional trading performance by the leading account.

### 3. Profit Concentration Among Top Traders
- A significant portion of total profits is concentrated within a small number of accounts.
- A few highly successful traders appear to drive a large share of overall profitability.

### 4. Third-Ranked Trader Earned Considerably Less
- The third trader generated 67,845.62 in total profit.
- There is a substantial drop in profitability after the top two traders.

### 5. Evidence of Uneven Trading Performance
- The wide gap between trader profits indicates that trading success is not evenly distributed across participants.
- Some traders consistently outperform the broader market.

### Overall Insight
- Trading profits are highly concentrated among a few top-performing accounts. The leading trader generated exceptional returns, significantly outperforming all other traders, highlighting the importance of trading strategy and execution quality in achieving superior results.

# Does volatility increase in Greed market?

In [112]:
volatility = (
    sentimental_analysis_df
    .groupby('classification')['Closed PnL']
    .std()
    .reset_index()
)



In [113]:
volatility

,classification,Closed PnL
0,Extreme Fear,2508.465987
1,Extreme Greed,175.430976
2,Fear,599.960527
3,Greed,432.521100
4,Neutral,661.095246


In [114]:
fig = px.bar(
    volatility,
    x="classification",
    y="Closed PnL",
    title="Volatility (PnL Std Dev) by Sentiment",
    text="Closed PnL",
    color="Closed PnL",
    color_continuous_scale="Reds"
)

fig.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig.show()

## Average Profit per Trader by Market Sentiment

### 1. Extreme Fear Generated the Highest Profit per Trader
- Traders earned an average profit of 2,508.47 during Extreme Fear.
- This is significantly higher than all other sentiment categories, suggesting strong profit opportunities during market panic.

### 2. Neutral Sentiment Delivered the Second-Highest Profit
- Neutral sentiment recorded an average profit of 661.10 per trader.
- Stable market conditions appear to support consistent trader performance.

### 3. Fear Outperformed Greed
- Fear generated an average profit of 599.96 per trader, compared to 432.52 during Greed.
- Traders were generally more profitable in cautious markets than in optimistic markets.

### 4. Extreme Greed Produced the Lowest Profit
- Extreme Greed recorded only 175.43 average profit per trader.
- Excessive market optimism appears to be associated with weaker trading performance.

### 5. Market Panic Created the Best Trading Opportunities
- The average profit during Extreme Fear is nearly four times higher than during Neutral sentiment and more than fourteen times higher than during Extreme Greed.
- This highlights the potential advantage of trading during periods of market stress.

### Overall Insight
- Traders achieved the highest average profits during Extreme Fear, while Extreme Greed produced the weakest results. The findings suggest that fearful market conditions may offer better trading opportunities than highly optimistic market environments.

# Which direction (Buy/Sell) is more profitable overall?

In [115]:
direction_profit = (
    sentimental_analysis_df
    .groupby('Side')['Closed PnL']
    .sum()
    .reset_index()
)



In [117]:
direction_profit

,Side,Closed PnL
0,BUY,330700.551708
1,SELL,877562.828194


In [116]:
fig = px.pie(
    direction_profit,
    names="Side",
    values="Closed PnL",
    title="Buy vs Sell Total Profit Contribution"
)

fig.show()

## Profitability by Trade Side

### 1. SELL Trades Generated the Highest Total Profit
- SELL trades produced a total Closed PnL of 877,562.83.
- This is significantly higher than the profit generated from BUY trades.

### 2. BUY Trades Were Less Profitable
- BUY trades generated a total Closed PnL of 330,700.55.
- While profitable, their contribution was much lower compared to SELL trades.

### 3. SELL Trades Outperformed BUY Trades by a Large Margin
- Total profit from SELL trades was more than 2.5 times higher than BUY trades.
- This indicates that SELL positions were more successful during the analyzed period.

### 4. Market Conditions Favored SELL Strategies
- The strong performance of SELL trades suggests that traders benefited more from short-selling or profit-taking strategies.
- Market movements during the period may have created more opportunities for SELL positions.

### Overall Insight
- SELL trades were the primary driver of trader profitability, generating the majority of total profits. This suggests that adapting to bearish market conditions or effectively managing exits played an important role in achieving higher returns.

# Sentiment vs Profit Relationship

In [123]:
avg_pnl = sentimental_analysis_df.groupby('classification')['Closed PnL'].mean().reset_index()
avg_pnl.columns = ['classification', 'avg_pnl']

In [124]:
avg_pnl

,classification,avg_pnl
0,Extreme Fear,198.456185
1,Extreme Greed,33.592341
2,Fear,79.542449
3,Greed,58.085638
4,Neutral,136.084118


In [125]:

fig = px.bar(
    avg_pnl,
    x="classification",
    y="avg_pnl",
    title="Average Profit by Market Sentiment",
    labels={
        "classification":"Sentiment",
        "avg_pnl":"Average Profit"
    },
    text="avg_pnl",
    color="avg_pnl",
    color_continuous_scale="Blues"
)

fig.update_traces(
    texttemplate='%{text:.2f}',
    textposition='outside'
)

fig.show()

## Average Profit (PnL) by Market Sentiment

### 1. Extreme Fear Generated the Highest Average Profit
- Extreme Fear recorded the highest average PnL of 198.46.
- Traders achieved their best performance during periods of extreme market pessimism.

### 2. Neutral Sentiment Delivered Strong Returns
- Neutral sentiment generated an average PnL of 136.08.
- Stable market conditions appear to support consistent profitability.

### 3. Fear Outperformed Greed
- Fear recorded an average PnL of 79.54, compared to 58.09 during Greed.
- Traders were generally more profitable in fearful markets than in optimistic markets.

### 4. Extreme Greed Produced the Lowest Average Profit
- Extreme Greed recorded the lowest average PnL of 33.59.
- Excessive market optimism appears to be associated with weaker trading performance.

### 5. Profitability Declines as Market Optimism Increases
- Average profits are highest during Extreme Fear and lowest during Extreme Greed.
- This suggests that traders may find better opportunities when markets are driven by fear rather than greed.

### Overall Insight
- Trader profitability was strongest during Extreme Fear and Neutral market conditions, while Extreme Greed delivered the weakest results. The findings indicate that cautious or panic-driven markets may offer more profitable trading opportunities than highly optimistic market environments.

# Sentiment vs Risk (Leverage Behavior)

In [120]:
avg_leverage = sentimental_analysis_df.groupby('classification')['Start Position'].mean().reset_index()
avg_leverage.columns = ['classification', 'avg_leverage']

In [121]:
avg_leverage

,classification,avg_leverage
0,Extreme Fear,16990.582744
1,Extreme Greed,2398.942883
2,Fear,5416.268693
3,Greed,11303.630871
4,Neutral,8184.226737


In [126]:
fig = px.bar(
    avg_leverage,
    x="classification",
    y="avg_leverage",
    title="Average Leverage by Market Sentiment",
    labels={
        "classification":"Sentiment",
        "avg_leverage":"Average Leverage"
    },
    text="avg_leverage",
    color="avg_leverage",
    color_continuous_scale="Reds"
)

fig.update_traces(
    texttemplate='%{text:.2f}',
    textposition='outside'
)

fig.show()

## Average Leverage (Start Position) by Market Sentiment

### 1. Extreme Fear Has the Highest Average Leverage
- Extreme Fear recorded the highest average leverage of 16,990.58.
- Traders took the largest positions during periods of extreme market fear.

### 2. Greed Shows High Leverage Usage
- Greed recorded an average leverage of 11,303.63.
- Traders appeared more confident and increased their exposure during optimistic market conditions.

### 3. Neutral Markets Maintain Moderate Leverage
- Neutral sentiment recorded an average leverage of 8,184.23.
- Position sizes remained relatively balanced in stable market conditions.

### 4. Fear Has Lower Leverage Than Greed
- Fear recorded an average leverage of 5,416.27.
- Traders were more cautious with position sizing compared to Greed and Extreme Fear.

### 5. Extreme Greed Has the Lowest Average Leverage
- Extreme Greed recorded the lowest average leverage of 2,398.94.
- Despite strong market optimism, traders used smaller positions on average.

### Overall Insight
- Traders used the highest leverage during Extreme Fear and the lowest during Extreme Greed. This suggests that market panic encouraged larger position sizes, while excessive optimism was associated with more conservative leverage usage.

# Sentiment vs Win Rate (Consistency)

In [118]:
sentimental_analysis_df['Win'] = (sentimental_analysis_df['Closed PnL'] > 0).astype(int)

win_rate = sentimental_analysis_df.groupby('classification')['Win'].mean().reset_index()
win_rate['win_rate'] = win_rate['Win'] * 100

In [119]:
win_rate

,classification,Win,win_rate
0,Extreme Fear,0.315301,31.530139
1,Extreme Greed,0.165625,16.562500
2,Fear,0.388217,38.821729
3,Greed,0.347269,34.726940
4,Neutral,0.498709,49.870912


In [128]:
fig = px.bar(
    win_rate,
    x="classification",
    y="win_rate",
    title="Win Rate by Market Sentiment",
    labels={
        "classification":"Sentiment",
        "win_rate":"Win Rate (%)"
    },
    text="win_rate",
    color="win_rate",
    color_continuous_scale="Greens"
)

fig.update_traces(
    texttemplate='%{text:.2f}%',
    textposition='outside'
)

fig.show()

## Win Rate by Market Sentiment

### 1. Neutral Sentiment Has the Highest Win Rate
- Neutral sentiment achieved the highest win rate of 49.87%.
- Nearly half of all trades were profitable during stable market conditions.

### 2. Fear Recorded the Second-Highest Win Rate
- Fear achieved a win rate of 38.82%.
- Traders performed relatively well even when market sentiment was negative.

### 3. Greed Shows Moderate Trading Success
- Greed recorded a win rate of 34.73%.
- Trader performance was lower than Neutral and Fear conditions.

### 4. Extreme Fear Outperformed Extreme Greed
- Extreme Fear achieved a win rate of 31.53%.
- Extreme Greed achieved only 16.56%, the lowest among all sentiment categories.
- Traders were nearly twice as successful during Extreme Fear compared to Extreme Greed.

### 5. Excessive Optimism Led to Poorer Performance
- The lowest win rate was observed during Extreme Greed.
- This may indicate that overconfidence and aggressive market behavior reduce trading success.

### Overall Insight
- Traders achieved the highest success rate during Neutral market conditions, followed by Fear. In contrast, Extreme Greed produced the lowest win rate, suggesting that stable and cautious markets are generally more favorable for profitable trading than highly optimistic markets.